### PR TITLE
feat: Finish implementing project creation on backend (#21)

### PR DESCRIPTION
**Pull request by ermshiperete**
This change finishes the implementation of creating a project on the backend (#21).

This change also implements extracting the keyboard id from the keyboard_info file (part of #10). 

---
This change is [Reviewable](https://reviewable.io/reviews/keymanapp/developer.keyman.com/38)

### CODE REVIEW

File: test/projects.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  let app: INestApplication;
+  let httpService: HttpService;
+  let token: string;
+  let user: string;
+  let testSession: any;
+  let authenticatedSession: any;
+  let githubService: GithubService;
+
+  async function deleteKeyboardsRepo(): Promise<void> {
+    await httpService.delete(`https://api.github.com/repos/${user}/keyboards`, {
```
Comment (mcdurdin) [inline test/projects.e2e-spec.ts]: This seems like a dangerous test. Do we need to reinforce that this test must not be run with the developer's real GH account?

File: README.md [hunk 2]
```diff
@@ -32,6 +32,15 @@ In the `*.env` file replace the values for `CLIENT_ID` and `CLIENT_SECRET` with
 _Client Secret_ that GitHub displays for the app. You should also replace the value for
 `SESSION_SECRET` with a random value.
 
+If you want to run all e2e tests, you'll have to create a test user on GitHub. Fork
+[kdotester1/khmer_angkor](https://github.com/kdotester1/khmer_angkor) to your test
```
[the remark this answers is not in the document]
Comment (ermshiperete) [reply README.md]: Good idea. Can you fork `kdotester1/test_kdo_khmer_angkor` and `keymanapptest/test_kdo_keyboards` into @keymanapp?

File: test/projects.e2e-spec.ts (hunk 1 above)
Comment (ermshiperete) [reply test/projects.e2e-spec.ts]: With the `keyboards` repo renamed `test_kdo_keyboards` this should be save now.

**Comment by ermshiperete** [comment]
> I'd like to see some "why" or "reason for existence" comments on some of the functions because there's a lot of code and a bit of context can really help understanding :)

Can you point out the methods where this would be good?

**Comment by mcdurdin** [comment]
> Can you point out the methods where this would be good?

For example, `waitForRepoToExist`. I can see what the code is doing but some context would be helpful. I like to see a one line description of the function because when someone else comes into the code, the 'why' is the hardest part. The code you've written is generally clear and easy to understand so there's no issue with the implementation (although I have to ask if it is right that `waitForRepoToExist` function really could ping GitHub 300 times in 5 minutes? That could cause trouble with rate limiting? Perhaps back-off might be advisable...)

**Comment by mcdurdin** [comment]
Okay, I've done the review but currently having trouble getting the test environment to run on my machine. Not sure if I am missing something in the setup (noting also that README.md between the root and frontend folders are perhaps slightly inconsistent)

**Comment by mcdurdin** [comment]
Okay, got the dev environment running. It seems to be mostly working, so far.

# Testing notes

## Workflow issue if forking repo late

1. I forked the repo after loading the KDO home page. This led to an issue where KDO did not show my keyboard. Instead, I was shown a spinner with a link "I don't see my keyboard project". 

image

2. I then forked the repo:

image

3. and clicked the link. This led to a blank screen (just the top banner visible):

image

4. The second time I tried to repro this, I got the following:

image

This may correspond to a warning in the log:

```
(node:17644) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 disconnect listeners added to [MemoryStore]. Use emitter.setMaxListeners() to increase limit
(node:17644) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 connect listeners added to [MemoryStore]. Use emitter.setMaxListeners() to increase limit
```

5. So refreshing the page didn't help; I had to logout and login again, to see the repo:

image

6. Once we get here, I click on the link and am presented with a blank screen but then I think that's as far as it goes in this PR, right?

## UX

It may be a little early to be worrying about this but the grid layout with small boxes for repo names (see above) are a little awkward. May be better to use a list layout for this.

**Comment by ermshiperete** [comment]
>     Once we get here, I click on the link and am presented with a blank screen but then I think that's as far as it goes in this PR, right?

yes, this PR only changes things in the backend, so everything you experienced is expected (with the exception of the "Too many requests" (#41)).

> It may be a little early to be worrying about this but the grid layout with small boxes for repo names (see above) are a little awkward. May be better to use a list layout for this.

I created #40 for that.

**Pull request merged**